---
layout: post
title:  "Promise小记"
date:   2019-10-01 08:00:00 +0800
categories: javascript
tag: [javascript]
---

Promise的定义


一个构造函数

    Promise(onFulfilled, onRejected)

两个方法：

    then(onFulfilled, onRejected)
    
    catch(onRejected)

其中 catch 是 then 的快捷方式，等效关系如下：

    catch(onRejected) == then(null, onRejected)

三条原则：

1. then 和 catch 的返回值是一个新的 Promise
2. Promise 的流式调用
3. 只有 throw 和 Promise.reject() 会触发 onRejected

<!-- more -->

几个典型的例子：

1. then 和 catch 的返回值是一个新的 Promise

2. rejected 的 Promise 会触发 onRejected

```javascript
    Promise.reject({})
        .then(()=>{
            console.log('ok'); //不会调用
        })
        .catch(()=>{
            console.log('error'); // print 'error'
        });

    Promise.reject({})
        .then(
            ()=>{
                console.log('ok'); //不会调用
            },
            ()=>{
              console.log('error'); // print 'error'
            }
        );
```

3. 只有 throw 和 Promise.reject() 会触发 onRejected

```javascript
    Promise.resolve({})
        .then(()=>{
            return new Error();
        })
        .then(()=>{
            console.log('ok'); // print 'ok'
        })
        .catch(()=>{
            console.log('error'); //不会调用
        });

    Promise.resolve({})
        .then(()=>{
            throw new Error();
        })
        .then(()=>{
            console.log('ok'); //不会调用
        })
        .catch(()=>{
            console.log('error'); // print 'error'
        });

    Promise.resolve({})
        .then(()=>{
            return Promise.reject({});
        })
        .then(()=>{
            console.log('ok'); //不会调用
        })
        .catch(()=>{
            console.log('error'); // print 'error'
        });
```


Promise 的滥用

